TokboxPublisher.js

```diff
@@ -0,0 +1,208 @@
+  const { audioSource, videoSource } = getPublisherSources()
+  return {
+    ...publisherOptions,
+    audioSource: (micEnabled || smart) ? audioSource : false,
+    videoSource: (camEnabled || smart) ? videoSource : false,
+    publishAudio: micEnabled,
+    publishVideo: camEnabled,
+    name: name + (lastname ? ' ' + lastname : '')
+  }
+}
+
+function getPublisherSources () {
+  const { micDevID, webcamDevID } = JSON.parse(localStorage.iosettings || null) || {}
+  const { isWebsiteHasWebcamPermissions, isWebsiteHasMicrophonePermissions, hasMicrophone, hasWebcam } = DetectRTC
+  let theWebcamDevID = hasWebcam 
+  let theMicDevID = hasMicrophone
+  if (isWebsiteHasWebcamPermissions && webcamDevID) theWebcamDevID = webcamDevID
+  if (isWebsiteHasMicrophonePermissions && micDevID) theMicDevID = micDevID
+  
+  return {
+    audioSource: theMicDevID,
+    videoSource: theWebcamDevID
+  }
+}
+
+function destroyPublisher () {
+  if (!publisher) return
+  destroyAutoplayStream(publisher.streamId)
+  publisher.destroy()
+}
+
+function onInitPublisherError ({ error, error: { name } = {}, options: { videoSource, audioSource } }) {  
+  if (['OT_CONSTRAINTS_NOT_SATISFIED', 'OT_HARDWARE_UNAVAILABLE'].includes(name)) {
+    // TODO: get rid of notifyDialog emitLocalEvent('showErrorNotification', 'Establishing mic stream failed...')
+    DEMIO.ui.notifyDialog(
+      'Establishing device stream failed. Ensure the device is properly connected and correctly selected in the Demio settings and try again.'
+    )
+  }
+  if (["OT_USER_MEDIA_ACCESS_DENIED", "OT_NO_DEVICES_FOUND"].includes(name)) {
+    emitLocalEvent('showInfoNoPermissions', { videoSource, audioSource })
+  }
+  if (["OT_HARDWARE_UNAVAILABLE", "OT_CHROME_MICROPHONE_ACQUISITION_ERROR"].includes(name)) {
+    emitLocalEvent('showNotReadableError')
+  }
+  if ("OT_CHROME_MICROPHONE_ACQUISITION_ERROR" === name) {
+    emitLocalEvent('showNotReadableError')
+  }
+
+  // TODO: Retry to publish???
+  emitLocalEvent('TokboxSession:initPublisher:error', { error, videoType: videoSource })
+  if ("OT_USER_MEDIA_ACCESS_DENIED" !== name) {
+    Raven.captureException(error)
+  }
+  log('initPublisher error', error)
+}
+
+function createPublisher ({ micEnabled, camEnabled, name, lastname, userID }, { smart } = {}) {
+  const options = getPublisherOptions({ micEnabled, camEnabled, name, lastname }, { smart })
+  let myPublisher = OpenTok.initPublisher(options, (error) => error && onInitPublisherError({ error, options }))
+
+  // State of the mic & cam to publish using the same instance
+  myPublisher.demioMicEnabled = micEnabled
+  myPublisher.demioCamEnabled = camEnabled
+  
+  let audioLevel = 0
+
+  myPublisher.on({
+    videoElementCreated: ({ element }) => {
+      myPublisher.demioVideoElement = element
+      element.onloadedmetadata = function () {
+        // Hack for autoplay when MediaStream is updated when the cam is enabled after the mic
```

Comment:
Can you elaborate please?

Reply:
We use an auto play hack that allows us to play video with audio unmuted because we do not use the video element created by the SDK. Instead we use our own video elements created right after the user clicks on the join button. When the SDK creates the video we simply reassign the mediaStream from one video to other using the attribute srcObject. 

We found that when the publisher is only publishing audio and then the webcam is enabled the SDK creates a different mediaStream and breaks the reference to our hacked video element. The solution that we found is detect when a new mediaStream is loaded on the SDK video to connect it with our hacked video element.

Reply:
ok makes sense :) Thank you